index.js

```diff
@@ -1051,40 +1061,46 @@
       };
     };
 
-    //  _determineActualTypes :: Boolean -> [Any] -> [Type]
-    var _determineActualTypes = function(loose) {
-      return function recur(values) {
-        if (isEmpty(values)) return [Unknown];
-        //  consistentTypes :: [Type]
-        var consistentTypes = chain(env, rejectAny);
-        //  typeses :: [[Type]]
-        var typeses = map(values, function(value) {
-          return chain(consistentTypes, function(t) {
-            return (
-              t.name === 'sanctuary-def/Nullable' || !test(t, value).valid ?
-                [] :
-              t.type === 'UNARY' ?
-                map(recur(t._1(value)), UnaryType.from(t)) :
-              t.type === 'BINARY' ?
-                BinaryType.xprod(t, recur(t._1(value)), recur(t._2(value))) :
-              // else
-                [t]
-            );
-          });
-        });
-        //  common :: [Type]
-        var common = commonTypes(typeses, loose);
-        if (!isEmpty(common)) return common;
-        //  If none of the values is a member of a type in the environment,
-        //  and all the values have the same type identifier, the values are
-        //  members of a "foreign" type.
-        if (isEmpty(filterTypesByValues(consistentTypes, values)) &&
-            all(values.slice(1), $$typeEq($$type(values[0])))) {
-          //  Create a nullary type for the foreign type.
-          return [type0($$type(values[0]))];
+    //  _determineActualTypes :: (Boolean, [Object], [Any]) -> [Type]
+    var _determineActualTypes = function recur(loose, $seen, values) {
+      if (isEmpty(values)) return [Unknown];
+      //  consistentTypes :: [Type]
+      var consistentTypes = chain(env, rejectAny);
+      //  typeses :: [[Type]]
+      var typeses = map(values, function(value) {
+        if (typeof value === 'object' && value != null) {
```

Comment:
Functions can have references too, and fail this check. Also Regex was considered a function in Chrome 1 to 12.